File: src/test/java/com/example/the_machine/service/ArtifactServiceTest.java
package com.example.the_machine.service;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import com.example.the_machine.common.IdGenerator;
import com.example.the_machine.domain.ArtifactEntity;
import com.example.the_machine.repo.ArtifactRepository;
import java.util.UUID;
import lombok.val;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

@ExtendWith(MockitoExtension.class)
class ArtifactServiceTest {

  @Mock
  private ArtifactRepository artifactRepository;

  @Mock
  private IdGenerator idGenerator;

  @InjectMocks
  private ArtifactService artifactService;

  @Captor
  private ArgumentCaptor<ArtifactEntity> artifactCaptor;

  @Test
  void shouldCreateStructuredJsonArtifact() {
    // Given
    val threadId = UUID.randomUUID();
    val runId = UUID.randomUUID();
    val artifactId = UUID.randomUUID();

    when(idGenerator.newId()).thenReturn(artifactId);
    when(artifactRepository.save(any(ArtifactEntity.class))).thenAnswer(
        invocation -> invocation.getArgument(0));

    // When
    val result = artifactService.createArtifact(
        threadId, runId,
        ArtifactEntity.Kind.STRUCTURED_JSON,
        ArtifactEntity.Stage.SCRIPT,
        "Apple Marketing Pitch",
        null
    );

    // Then
    verify(artifactRepository).save(artifactCaptor.capture());
    val savedArtifact = artifactCaptor.getValue();

    assertThat(savedArtifact.getId()).isEqualTo(artifactId);
    assertThat(savedArtifact.getThreadId()).isEqualTo(threadId);
    assertThat(savedArtifact.getRunId()).isEqualTo(runId);
    assertThat(savedArtifact.getKind()).isEqualTo(ArtifactEntity.Kind.STRUCTURED_JSON);
    assertThat(savedArtifact.getStage()).isEqualTo(ArtifactEntity.Stage.SCRIPT);
    assertThat(savedArtifact.getStatus()).isEqualTo(ArtifactEntity.Status.PENDING);
    assertThat(savedArtifact.getTitle()).isEqualTo("Apple Marketing Pitch");
    assertThat(savedArtifact.getContent()).isNull(); // No content initially
    assertThat(savedArtifact.getCreatedAt()).isNotNull();
    assertThat(result).isEqualTo(savedArtifact);
  }

  @Test
  void shouldCreateImageArtifact() {
    // Given
    val threadId = UUID.randomUUID();
    val runId = UUID.randomUUID();
    val artifactId = UUID.randomUUID();

    when(idGenerator.newId()).thenReturn(artifactId);
    when(artifactRepository.save(any(ArtifactEntity.class))).thenAnswer(
        invocation -> invocation.getArgument(0));

    // When
    val result = artifactService.createArtifact(
        threadId, runId,
        ArtifactEntity.Kind.IMAGE,
        ArtifactEntity.Stage.IMAGES,
        "Marketing Visual 1",
        "image/png"
    );

    // Then
    verify(artifactRepository).save(artifactCaptor.capture());
    val savedArtifact = artifactCaptor.getValue();

    assertThat(savedArtifact.getKind()).isEqualTo(ArtifactEntity.Kind.IMAGE);
    assertThat(savedArtifact.getStage()).isEqualTo(ArtifactEntity.Stage.IMAGES);
    assertThat(savedArtifact.getTitle()).isEqualTo("Marketing Visual 1");
    assertThat(savedArtifact.getMime()).isEqualTo("image/png");
    assertThat(savedArtifact.getStorageKey()).isNull(); // No storage key initially
    assertThat(savedArtifact.getContent()).isNull(); // Images don't have inline content
    assertThat(result).isEqualTo(savedArtifact);
  }

  @Test
  void shouldCreatePdfArtifact() {
    // Given
    val threadId = UUID.randomUUID();
    val runId = UUID.randomUUID();
    val artifactId = UUID.randomUUID();

    when(idGenerator.newId()).thenReturn(artifactId);
    when(artifactRepository.save(any(ArtifactEntity.class))).thenAnswer(
        invocation -> invocation.getArgument(0));

    // When
    val result = artifactService.createArtifact(
        threadId, runId,
        ArtifactEntity.Kind.PDF,
        ArtifactEntity.Stage.DECK,
        "Apple Marketing Pitch - Revised",
        "application/pdf"
    );

    // Then
    verify(artifactRepository).save(artifactCaptor.capture());
    val savedArtifact = artifactCaptor.getValue();

    assertThat(savedArtifact.getKind()).isEqualTo(ArtifactEntity.Kind.PDF);
    assertThat(savedArtifact.getStage()).isEqualTo(ArtifactEntity.Stage.DECK);
    assertThat(savedArtifact.getTitle()).isEqualTo("Apple Marketing Pitch - Revised");
    assertThat(savedArtifact.getMime()).isEqualTo("application/pdf");
    assertThat(savedArtifact.getStorageKey()).isNull(); // No storage key initially
    assertThat(result).isEqualTo(savedArtifact);
  }
}

File: src/test/kotlin/com/example/the_machine/service/ArtifactServiceTest.kt
package com.example.the_machine.service

import com.example.the_machine.common.UuidV7
import com.example.the_machine.config.TestDataConfiguration.TEST_ARTIFACT_ID
import com.example.the_machine.config.TestDataConfiguration.TEST_RUN_ID
import com.example.the_machine.config.TestDataConfiguration.TEST_THREAD_ID
import com.example.the_machine.config.TestDataConfiguration.TEST_THREAD_ID_2
import com.example.the_machine.config.TestDataConfiguration.TEST_THREAD_ID_3
import com.example.the_machine.domain.ArtifactEntity
import com.example.the_machine.repo.ArtifactRepository
import io.mockk.*
import io.mockk.junit5.MockKExtension
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith

@ExtendWith(MockKExtension::class)
class ArtifactServiceTest {

  private val artifactRepository: ArtifactRepository = mockk()
  private val artifactService = ArtifactService(artifactRepository)
  private val artifactSlot = slot<ArtifactEntity>()

  @Test
  fun shouldCreateStructuredJsonArtifact() {
    mockkObject(UuidV7)
    try {
      // Given
      val threadId = TEST_THREAD_ID
      val runId = TEST_RUN_ID

      every { UuidV7.newId() } returns TEST_ARTIFACT_ID
      every { artifactRepository.save(capture(artifactSlot)) } answers { artifactSlot.captured }

      // When
      val result = artifactService.createArtifact(
        threadId, runId,
        ArtifactEntity.Kind.STRUCTURED_JSON,
        ArtifactEntity.Stage.SCRIPT,
        "Apple Marketing Pitch",
        null
      )

      // Then
      verify { artifactRepository.save(any()) }
      val savedArtifact = artifactSlot.captured

      assertThat(savedArtifact.id).isEqualTo(TEST_ARTIFACT_ID)
      assertThat(savedArtifact.threadId).isEqualTo(threadId)
      assertThat(savedArtifact.runId).isEqualTo(runId)
      assertThat(savedArtifact.kind).isEqualTo(ArtifactEntity.Kind.STRUCTURED_JSON)
      assertThat(savedArtifact.stage).isEqualTo(ArtifactEntity.Stage.SCRIPT)
      assertThat(savedArtifact.status).isEqualTo(ArtifactEntity.Status.PENDING)
      assertThat(savedArtifact.title).isEqualTo("Apple Marketing Pitch")
      assertThat(savedArtifact.content).isNull() // No content initially
      assertThat(result).isEqualTo(savedArtifact)
    } finally {
      unmockkObject(UuidV7)
    }
  }

  @Test
  fun shouldCreateImageArtifact() {
    mockkObject(UuidV7)
    try {
      // Given
      val threadId = TEST_THREAD_ID_2
      val runId = TEST_RUN_ID

      every { UuidV7.newId() } returns TEST_ARTIFACT_ID
      every { artifactRepository.save(capture(artifactSlot)) } answers { artifactSlot.captured }

      // When
      val result = artifactService.createArtifact(
        threadId, runId,
        ArtifactEntity.Kind.IMAGE,
        ArtifactEntity.Stage.IMAGES,
        "Marketing Visual 1",
        "image/png"
      )

      // Then
      verify { artifactRepository.save(any()) }
      val savedArtifact = artifactSlot.captured

      assertThat(savedArtifact.id).isEqualTo(TEST_ARTIFACT_ID)
      assertThat(savedArtifact.kind).isEqualTo(ArtifactEntity.Kind.IMAGE)
      assertThat(savedArtifact.stage).isEqualTo(ArtifactEntity.Stage.IMAGES)
      assertThat(savedArtifact.title).isEqualTo("Marketing Visual 1")
      assertThat(savedArtifact.mime).isEqualTo("image/png")
      assertThat(savedArtifact.storageKey).isNull() // No storage key initially
      assertThat(savedArtifact.content).isNull() // Images don't have inline content
      assertThat(result).isEqualTo(savedArtifact)
    } finally {
      unmockkObject(UuidV7)
    }
  }

  @Test
  fun shouldCreatePdfArtifact() {
    mockkObject(UuidV7)
    try {
      // Given
      val threadId = TEST_THREAD_ID_3
      val runId = TEST_RUN_ID
      val artifactId = TEST_ARTIFACT_ID

      every { UuidV7.newId() } returns artifactId
      every { artifactRepository.save(capture(artifactSlot)) } answers { artifactSlot.captured }

      // When
      val result = artifactService.createArtifact(
        threadId, runId,
        ArtifactEntity.Kind.PDF,
        ArtifactEntity.Stage.DECK,
        "Apple Marketing Pitch - Revised",
        "application/pdf"
      )

      // Then
      verify { artifactRepository.save(any()) }
      val savedArtifact = artifactSlot.captured

      assertThat(savedArtifact.id).isEqualTo(artifactId)
      assertThat(savedArtifact.kind).isEqualTo(ArtifactEntity.Kind.PDF)
      assertThat(savedArtifact.stage).isEqualTo(ArtifactEntity.Stage.DECK)
      assertThat(savedArtifact.title).isEqualTo("Apple Marketing Pitch - Revised")
      assertThat(savedArtifact.mime).isEqualTo("application/pdf")
      assertThat(savedArtifact.storageKey).isNull() // No storage key initially
      assertThat(result).isEqualTo(savedArtifact)
    } finally {
      unmockkObject(UuidV7)
    }
  }
}